File: clients/android/app/src/main/kotlin/com/lextures/android/core/design/Haptics.kt
package com.lextures.android.core.design

import android.content.Context
import android.os.Build
import android.view.HapticFeedbackConstants
import android.view.View

/**
 * AN.6 — Standardized haptics mapping for control interactions (FR-5).
 *
 * Views perform feedback via [View.performHapticFeedback]; the OS haptics
 * setting is honored automatically. Never gate the control action on this (FR-9).
 */
object Haptics {
    enum class Kind {
        Tap,
        Selection,
        Success,
        Error,
    }

    /** System constant name for tests / logging. */
    fun systemName(kind: Kind): String = when (kind) {
        Kind.Tap -> "CONTEXT_CLICK"
        Kind.Selection -> "CLOCK_TICK"
        Kind.Success -> "CONFIRM"
        Kind.Error -> "REJECT"
    }

    fun feedbackConstant(kind: Kind): Int = when (kind) {
        Kind.Tap -> HapticFeedbackConstants.CONTEXT_CLICK
        Kind.Selection -> HapticFeedbackConstants.CLOCK_TICK
        Kind.Success -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            HapticFeedbackConstants.CONFIRM
        } else {
            HapticFeedbackConstants.CONTEXT_CLICK
        }
        Kind.Error -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            HapticFeedbackConstants.REJECT
        } else {
            HapticFeedbackConstants.LONG_PRESS
        }
    }

    fun trigger(view: View?, kind: Kind, enabled: Boolean = true) {
        if (!enabled || view == null) return
        view.performHapticFeedback(feedbackConstant(kind))
    }

    /** Pure helper for unit tests without a View. */
    fun shouldFire(enabled: Boolean, systemHapticsOff: Boolean = false): Boolean {
        if (!enabled) return false
        if (systemHapticsOff) return false
        return true
    }

    @Suppress("UNUSED_PARAMETER")
    fun isSystemHapticsOff(context: Context): Boolean {
        // View.performHapticFeedback already respects the user setting; this
        // helper exists for testability / explicit checks.
        return false
    }
}

File: clients/ios/Lextures/Core/Design/Haptics.swift
import UIKit

/// AN.6 — Standardized haptics mapping for control interactions (FR-5).
///
/// Respects the OS haptics / reduce-motion settings via `UIFeedbackGenerator`.
/// Never call from the main action path in a way that gates the handler (FR-9).
enum Haptics {
    enum Kind: String, CaseIterable {
        case tap
        case selection
        case success
        case error
    }

    /// System feedback type name used for tests / logging (AC mapping table).
    static func systemName(for kind: Kind) -> String {
        switch kind {
        case .tap: return "lightImpact"
        case .selection: return "selection"
        case .success: return "notificationSuccess"
        case .error: return "notificationError"
        }
    }

    /// Fire a haptic if `enabled` and the OS allows feedback.
    static func trigger(_ kind: Kind, enabled: Bool = true) {
        guard enabled else { return }
        switch kind {
        case .tap:
            let gen = UIImpactFeedbackGenerator(style: .light)
            gen.prepare()
            gen.impactOccurred()
        case .selection:
            let gen = UISelectionFeedbackGenerator()
            gen.prepare()
            gen.selectionChanged()
        case .success:
            let gen = UINotificationFeedbackGenerator()
            gen.prepare()
            gen.notificationOccurred(.success)
        case .error:
            let gen = UINotificationFeedbackGenerator()
            gen.prepare()
            gen.notificationOccurred(.error)
        }
    }
}
